src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,491 @@
+
+    file_opened = 1;
+}
+
+/* ----------------------------------------------------------------------
+   create dynamical matrix
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::calculateMatrix()
+{
+    int local_idx; // local index
+    int local_jdx; // second local index
+    int nlocal = atom->nlocal;
+    bigint natoms = atom->natoms;
+    int *mask = atom->mask;
+    int *type = atom->type;
+    double imass; // dynamical matrix element
+    double *m = atom->mass;
+    double **f = atom->f;
+
+    //initialize dynmat to all zeros
+    for (int i=0; i < dynlen; i++)
+        for (int j=0; j < dynlen; j++)
+            dynmat[i][j] = 0.;
+
+    energy_force(0);
+
+    if (comm->me == 0 && screen) fprintf(screen,"Calculating Dynamical Matrix...\n");
+
+    for (int i=1; i<=natoms; i++){
+        local_idx = atom->map(i);
+        //if (screen) fprintf(screen, "local idx = %i on proc %i with %i local\n",local_idx, comm->me, nlocal);
+        if (local_idx >= 0){
+            for (int alpha=0; alpha<3; alpha++){
+                displace_atom(local_idx, alpha, 1);
+                energy_force(0);
+                for (int j=1; j<=natoms; j++){
+                    local_jdx = atom->map(j);
+                    if (local_jdx >= 0  && local_idx < nlocal){
+                        for (int beta=0; beta<3; beta++){
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] += -f[j-1][beta];
+                        }
+                    }
+                }
+                displace_atom(local_idx,alpha,-2);
+                energy_force(0);
+                for (int j=1; j<=natoms; j++){
+                    local_jdx = atom->map(j);
+                    if (local_jdx >= 0 && local_idx < nlocal){
+                        for (int beta=0; beta<3; beta++){
+                            if (atom->rmass_flag == 1)
+                                imass = sqrt(m[i] * m[j]);
+                            else
+                                imass = sqrt(m[type[i]] * m[type[j]]);
+                            //dynmat has length dynlen
+                            //dynlen = (group->count(igroup))*3;
+                            //currently dynmat is being called to natoms*3
+                            //Also with this implementation
+                            //I am not recovering the correct dynamical matrix
+                            //if I have more than 1 core
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] -= -f[j-1][beta];
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] /= (2 * del * imass);
+                            dynmat[(i-1)*3+alpha][(j-1)*3+beta] *= conversion;
+                        }
+                    }
+                }
+                displace_atom(local_idx,alpha,1);
+            }
+        }
+    }
+
+
+    memory->create(final_dynmat,int(dynlen),int(dynlen),"dynamic_matrix_buffer:buf");
```

Comment:
This limits `dynlen` to be a 32-bit only integer. Now, it is going to be quite slow to compute the dynamical matrix for a very large system, but then there should be a test at the beginning checking that `dynlen` isn't larger than `MAXSMALLINT`. Then it could also be made an `int` and loop indices correspondingly.